src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/KeywordService.kt

```diff
@@ -0,0 +1,270 @@
+    if (keywordRequest.prisonIds?.isEmpty() != false) {
+      log.warn("Invalid keyword search  - no prisonIds specified to filter by")
+      throw BadRequestException("Invalid keyword search  - please provide prison locations to filter by")
+    }
+  }
+
+  fun findByKeyword(keywordRequest: KeywordRequest): Page<Prisoner> {
+    log.info("Received keyword request ${gson.toJson(keywordRequest)}")
+
+    validateKeywordRequest(keywordRequest)
+    val searchSourceBuilder = createSourceBuilder(keywordRequest)
+    val searchRequest = SearchRequest(arrayOf(getIndex()), searchSourceBuilder)
+
+    // Useful for logging the JSON elastic search query that is executed
+    // log.info("Keyword query JSON: {}", searchSourceBuilder.toString())
+
+    return try {
+      val searchResponse = elasticSearchClient.search(searchRequest)
+      customEventForFindBySearchCriteria(keywordRequest, searchResponse.hits.totalHits?.value ?: 0)
+      createKeywordResponse(keywordRequest.pagination, searchResponse)
+    } catch (e: Throwable) {
+      log.error("Elastic search exception: $e")
+      createEmptyResponse(keywordRequest.pagination)
+    }
+  }
+
+  private fun createSourceBuilder(keywordRequest: KeywordRequest): SearchSourceBuilder {
+    val pageable = PageRequest.of(keywordRequest.pagination.page, keywordRequest.pagination.size)
+    return SearchSourceBuilder().apply {
+      timeout(TimeValue(searchTimeoutSeconds, TimeUnit.SECONDS))
+      size(pageable.pageSize.coerceAtMost(maxSearchResults))
+      from(pageable.offset.toInt())
+      sort("_score")
+      sort("prisonerNumber")
+      trackTotalHits(true)
+      query(buildKeywordQuery(keywordRequest))
+    }
+  }
+
+  private fun buildKeywordQuery(keywordRequest: KeywordRequest): BoolQueryBuilder {
+    val keywordQuery = QueryBuilders.boolQuery()
+
+    if (noKeyWordsSpecified(keywordRequest)) {
+      keywordQuery.should(QueryBuilders.matchAllQuery())
+    }
+
+    val sanitisedKeywordRequest = KeywordRequest(
+      orWords = addUppercaseKeywordTokens(keywordRequest.orWords),
+      andWords = addUppercaseKeywordTokens(keywordRequest.andWords),
+      exactPhrase = addUppercaseKeywordTokens(keywordRequest.exactPhrase),
+      notWords = addUppercaseKeywordTokens(keywordRequest.notWords),
+      prisonIds = keywordRequest.prisonIds,
+      fuzzyMatch = keywordRequest.fuzzyMatch ?: false,
+      pagination = keywordRequest.pagination,
+    )
+
+    with(sanitisedKeywordRequest) {
+
+      andWords.takeIf { !it.isNullOrBlank() }?.let {
+        // Will match only if all of these words are present across all text fields
+        keywordQuery.must(
+          QueryBuilders.boolQuery()
+            .minimumShouldMatch(1)
+            .should(
+              QueryBuilders.multiMatchQuery(it)
+                .type(MultiMatchQueryBuilder.Type.CROSS_FIELDS)
+                .lenient(true)
+                .fuzzyTranspositions(fuzzyMatch!!)
```

Comment:
minor, I guess this isn't nullable so no need for !! ?

Reply:
Think I altered this to default if null in the last round of testing, so will remove the !!.

Reply:
I checked this again and it is a nullable type `Boolean?` - which may not be provided by the client so I've defaulted it to false if it is ever null.